x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/ExponentialHistogramStates.java

```diff
@@ -184,6 +186,169 @@ public void close() {
+        }
+
+        public boolean isSeen() {
+            return histogramValue != null;
+        }
+
+        public long longValue() {
+            assert isSeen();
+            return longValue;
+        }
+
+        public ReleasableExponentialHistogram histogramValue() {
+            assert isSeen();
+            return histogramValue;
+        }
+
+        public void set(long longValue, ExponentialHistogram histogram) {
+            assert histogram != null;
+            this.longValue = longValue;
+            Releasables.close(histogramValue);
+            this.histogramValue = ExponentialHistogram.builder(histogram, new HistoBreaker(breaker)).build();
+        }
+
+        @Override
+        public void toIntermediate(Block[] blocks, int offset, DriverContext driverContext) {
+            assert blocks.length >= offset + 3;
+            BlockFactory blockFactory = driverContext.blockFactory();
+            // in case of error, the blocks are closed by the caller
+            if (histogramValue == null) {
+                blocks[offset] = blockFactory.newConstantLongBlockWith(0L, 1);
+                blocks[offset + 1] = blockFactory.newConstantExponentialHistogramBlock(ExponentialHistogram.empty(), 1);
+                blocks[offset + 2] = blockFactory.newConstantBooleanBlockWith(false, 1);
+            } else {
+                blocks[offset] = blockFactory.newConstantLongBlockWith(longValue, 1);
+                blocks[offset + 1] = blockFactory.newConstantExponentialHistogramBlock(histogramValue, 1);
+                blocks[offset + 2] = blockFactory.newConstantBooleanBlockWith(true, 1);
+            }
+        }
+
+        public Block evaluateFinalHistogram(DriverContext driverContext) {
+            BlockFactory blockFactory = driverContext.blockFactory();
+            if (histogramValue == null) {
+                return blockFactory.newConstantNullBlock(1);
+            } else {
+                return blockFactory.newConstantExponentialHistogramBlock(histogramValue, 1);
+            }
+        }
+
+        @Override
+        public void close() {
+            Releasables.close(histogramValue);
+            histogramValue = null;
+        }
+    }
+
+    /**
+     * A grouping state consisting of a single {@code long} value with a {@link ExponentialHistogram} per group.
+     * The intermediate state contains three values in order: the long, the histogram, and a boolean specifying if a value was set or not.
+     */
+    public static final class WithLongGroupingState implements GroupingAggregatorState {
+
+        private LongArray longValues;
+        private ObjectArray<ReleasableExponentialHistogram> histogramValues;
+        private final HistoBreaker breaker;
+        private final BigArrays bigArrays;
+
+        WithLongGroupingState(BigArrays bigArrays, CircuitBreaker breaker) {
+            this.longValues = bigArrays.newLongArray(1);
+            this.histogramValues = bigArrays.newObjectArray(1);
```

Comment:
We need to release `longValues` if we hit the circuit breaker while creating histogramValues.